nano/secure/ledger.cpp

```diff
@@ -466,12 +446,12 @@ void ledger_processor::change_block (nano::change_block & block_a)
 			result.code = block_a.valid_predecessor (*previous) ? nano::process_result::progress : nano::process_result::block_position;
 			if (result.code == nano::process_result::progress)
 			{
-				auto account (ledger.store.frontier.get (transaction, block_a.hashables.previous));
-				result.code = account.is_zero () ? nano::process_result::fork : nano::process_result::progress;
+				auto account = ledger.account (*previous);
```

Comment:
Can we trust the sideband to have the account information?
I noticed recently that (copied from slack chat):
```
The field account of class nano::block_sideband is always set to all zeros for state blocks and for open blocks read from the ledger. Is that a bug? It seems very strange to have a field that is mostly incorrect
```

Reply:
This is using ledger::account which will get it either from the sideband when appropriate or directly from the block

`nano::account nano::ledger::account (nano::block const & block) const
{
	debug_assert (block.has_sideband ());
	nano::account result (block.account ());
	if (result.is_zero ())
	{
		result = block.sideband ().account;
	}
	debug_assert (!result.is_zero ());
	return result;
}`

I do have some concern with this implementation as it's not obviously how it works, it returns the non-zero value from either sideband or the block, or zero if they're both zero. The way I have this implemented in the block checker context is more direct:

`nano::account nano::block_check_context::account () const
{
	switch (block->type ())
	{
		case nano::block_type::change:
		case nano::block_type::receive:
		case nano::block_type::send:
			debug_assert (previous != nullptr);
			switch (previous->type ())
			{
				case nano::block_type::state:
				case nano::block_type::open:
					return previous->account ();
				case nano::block_type::change:
				case nano::block_type::receive:
				case nano::block_type::send:
					return previous->sideband ().account;
				case nano::block_type::not_a_block:
				case nano::block_type::invalid:
					debug_assert (false);
					break;
			}
			break;
		case nano::block_type::state:
		case nano::block_type::open:
			return block->account ();
		case nano::block_type::not_a_block:
		case nano::block_type::invalid:
			debug_assert (false);
			break;
	}
	// std::unreachable (); c++23
	return 1; // Return an account that cannot be signed for.
}
`